accepted/2022/FunctionPointers.md

```diff
@@ -0,0 +1,547 @@
+
+This means that if we don't expose the calling conventions on `Type` for function pointers, some function pointer types will be castable and others not, and with no obvious way to inspect the `Type` to determine why. We could exposed just the "CallKind" byte on `Type`, but that is not desired since that will confuse and\or complicate the consumer's usage -- see "Normalization of calling conventions".
+
+Since we must already do the work of exposing the calling conventions including the newer ones that are read via custom modifiers, it makes sense then to expose all custom modifieres (via the new `Type.GetRequiredCustomModifiers()` and `Type.GetOptionalCustomModifiers()`.
+
+## Type identity 
+### Using of void*
+A function pointer addresscan be cast to any function pointer type when using `void*`:
+```cs
+delegate*<string> fn_string = &StringMethod;
+int i = ((delegate*<int>)(void*)fn_string)(); // Unsafe code allows any cast once void* is used
+static string StringMethod() => "Forty-Two";
+```
+And the IL just does a `ldftn` with no check:
+```
+.locals init (method string *() V_0, int32 V_1)
+ldftn      string ConsoleApp104.Program::'<Main>g__StringMethod|0_0'()
+stloc.0
+ldloc.0
+calli      int32()
+stloc.1
+```
+
+### Equality:
+```cs
+bool f = typeof(delegate*<string>) == typeof(delegate*<int>); // Expect 'false' (today 'true' since they are both IntPtr)
+```
+Note the use of `ldtoken`:
+```
+ldtoken    method string *()
+call       class [System.Runtime]System.Type [System.Runtime]System.Type::GetTypeFromHandle(valuetype [System.Runtime]System.RuntimeTypeHandle)
+ldtoken    method int32 *()
+call       class [System.Runtime]System.Type [System.Runtime]System.Type::GetTypeFromHandle(valuetype [System.Runtime]System.RuntimeTypeHandle)
+call       bool [System.Runtime]System.Type::op_Equality(class [System.Runtime]System.Type, class [System.Runtime]System.Type)
+```
+
+### Casting
+Using arrays to help:
+```cs
+object arrInt = new delegate*<int>[1];
+var yikes = (delegate*<string>[])arrInt; // Expect InvalidCastException
+```
+```
+.locals init (object V_0, method string *()[] V_1)
+ldc.i4.1
+newarr     method int32 *()
+stloc.0
+ldloc.0
+castclass  method string *()[]
+stloc.1
+```
+### `Is` keyword
+```cs
+object arrInt = new delegate*<int>[1];
+bool f = arrInt is delegate*<string>[]; // Expect 'false'
+```
+IL:
+```
+.locals init (object V_0, bool V_1)
+ldc.i4.1
+newarr     method int32 *()
+stloc.0
+ldloc.0
+isinst     method string *()[]
+ldnull
+cgt.un
+stloc.1
+```
+
+In all cases above* any calling conventions are included in type identity. For example,
+`delegate* unmanaged[Cdecl]<void>`
+and
+`delegate* unmanaged[FastCall]<void>`are not equal.
+
+_* not sure how the `is` (`isinst`) is implemented in the CLR and whether we can get the same behavior as `castclass`._
```

Comment:
The casting logic behind `isinst`, `castclass` and `Type.IsAssignableFrom` is shared.

Reply:
Thanks - removed that comment. However, I did add a section here for cases where C# hard-codes the equality check instead of using `isinst` for cases where it "knows" what the result is.